src/main/java/com/android/volley/cronet/CronetHttpStack.java

```diff
@@ -0,0 +1,284 @@
+                }
+                break;
+            case Request.Method.GET:
+                // Not necessary to set the request method because connection defaults to GET but
+                // being explicit here.
+                builder.setHttpMethod("GET");
+                break;
+            case Request.Method.DELETE:
+                builder.setHttpMethod("DELETE");
+                break;
+            case Request.Method.POST:
+                builder.setHttpMethod("POST");
+                addBodyIfExists(request.getBody(), builder);
+                break;
+            case Request.Method.PUT:
+                builder.setHttpMethod("PUT");
+                addBodyIfExists(request.getBody(), builder);
+                break;
+            case Request.Method.HEAD:
+                builder.setHttpMethod("HEAD");
+                break;
+            case Request.Method.OPTIONS:
+                builder.setHttpMethod("OPTIONS");
+                break;
+            case Request.Method.TRACE:
+                builder.setHttpMethod("TRACE");
+                break;
+            case Request.Method.PATCH:
+                builder.setHttpMethod("PATCH");
+                addBodyIfExists(request.getBody(), builder);
+                break;
+            default:
+                throw new IllegalStateException("Unknown method type.");
+        }
+    }
+
+    /**
+     * Sets the request headers for the UrlRequest.
+     *
+     * @param request to get the headers from
+     * @param additionalHeaders for the UrlRequest
+     * @param builder that we are adding the request headers to
+     * @throws AuthFailureError is thrown if Request#getHeaders throws ones
+     */
+    private void setRequestHeaders(
+            Request<?> request, Map<String, String> additionalHeaders, UrlRequest.Builder builder)
+            throws AuthFailureError {
+        HashMap<String, String> map = new HashMap<>();
+        map.putAll(additionalHeaders);
+        // Request.getHeaders() takes precedence over the given additional (cache) headers).
+        map.putAll(request.getHeaders());
+        for (Map.Entry<String, String> header : map.entrySet()) {
+            builder.addHeader(header.getKey(), header.getValue());
+        }
+    }
+
+    /** Sets the UploadDataProvider of the UrlRequest.Builder */
+    private void addBodyIfExists(@Nullable byte[] body, UrlRequest.Builder builder) {
+        if (body != null) {
+            UploadDataProvider dataProvider = UploadDataProviders.create(body);
+            builder.setUploadDataProvider(dataProvider, mCallbackExecutor);
+        }
+    }
+
+    /** Sets the priority of this request. */
+    private void setPriority(Request<?> request, UrlRequest.Builder builder) {
+        builder.setPriority(request.getPriority().ordinal());
+    }
+
+    private int getContentLength(UrlResponseInfo urlResponseInfo) {
+        List<String> content = urlResponseInfo.getAllHeaders().get("content-length");
```

Comment:
The capitalization should be "Content-Length", and it doesn't appear to me that this is a case-insensitive map (or guaranteed to be one), though most other places that look for headers appear to do so in a case-insensitive fashion:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Length
https://developer.android.com/guide/topics/connectivity/cronet/reference/org/chromium/net/UrlResponseInfo.html#getAllHeaders()

is it?

Reply:
hmm that's interesting. I printed all the headers and they were all lower case. And using "content-length" does seem to be working right now. I'll do some more testing with it.